src/main/java/emissary/core/SafeUsageChecker.java

```diff
@@ -0,0 +1,110 @@
+import java.util.Map;
+
+/**
+ * Utility for validating that Places safely interact with IBDO payloads in byte array form. Specifically, this class
+ * helps with validating that changes to a IBDO's payload are followed by a call to the
+ * {@link IBaseDataObject#setData(byte[]) setData(byte[])}, {@link IBaseDataObject#setData(byte[], int, int)
+ * setData(byte[], int, int)}, or {@link IBaseDataObject#setChannelFactory(SeekableByteChannelFactory)
+ * setChannelFactory(SeekableByteChannelFactory)} method.
+ */
+public class SafeUsageChecker {
+    protected static final Logger LOGGER = LoggerFactory.getLogger(SafeUsageChecker.class);
+
+    public static final String ENABLED_KEY = "ENABLED";
+    public static final boolean ENABLED_FROM_CONFIGURATION;
+    public static final String UNSAFE_MODIFICATION_DETECTED = "Detected unsafe changes to IBDO byte array contents";
+
+    static {
+        boolean enabledFromConfiguration = false;
+
+        try {
+            Configurator configurator = ConfigUtil.getConfigInfo(SafeUsageChecker.class);
+
+            enabledFromConfiguration = configurator.findBooleanEntry(ENABLED_KEY, enabledFromConfiguration);
+        } catch (IOException e) {
+            LOGGER.warn("Could not get configuration!", e);
+        }
+
+        ENABLED_FROM_CONFIGURATION = enabledFromConfiguration;
+    }
+
+    /**
+     * Cache that records each {@literal byte[]} reference made available to IBDO clients, along with a sha256 hash of the
+     * array contents. Used for determining whether the clients modify the array contents without explicitly pushing those
+     * changes back to the IBDO
+     */
+    private final Map<byte[], String> cache = new HashMap<>();
+    public final boolean enabled;
+
+    public SafeUsageChecker() {
+        enabled = ENABLED_FROM_CONFIGURATION;
+    }
+
+    public SafeUsageChecker(Configurator configurator) {
+        enabled = configurator.findBooleanEntry(ENABLED_KEY, ENABLED_FROM_CONFIGURATION);
+    }
+
+    /**
+     * Resets the snapshot cache
+     */
+    public void reset() {
+        if (enabled) {
+            cache.clear();
+        }
+    }
+
+    /**
+     * Stores a new integrity snapshot
+     * 
+     * @param bytes byte[] for which a snapshot should be captured
+     */
+    public void recordSnapshot(final byte[] bytes) {
+        if (enabled) {
+            cache.put(bytes, ByteUtil.sha256Bytes(bytes));
+        }
+    }
+
+
+    /**
+     * Resets the cache and stores a new integrity snapshot
+     * 
+     * @param bytes byte[] for which a snapshot should be captured
+     */
+    public void resetCacheThenRecordSnapshot(final byte[] bytes) {
+        if (enabled) {
+            reset();
+            recordSnapshot(bytes);
+        }
+    }
+
+    /**
+     * Uses the snapshot cache to determine whether any of the byte arrays have unsaved changes
+     * 
+     * @return boolean indication of unsafe changes
+     */
+    public boolean checkForUnsafeDataChanges() {
```

Comment:
Optional, but I think we should choose a method name that implies a `boolean` result.  `checkForUnsafeDataChanges` is a bit ambiguous.
```suggestion
    public boolean unsafeDataChangesDetected() {
```